src/tunnel.cc

```diff
@@ -310,7 +310,7 @@ TunnelStateData::serverClosed()
 {
     server.noteClosure();
 
-    request->hier.stopPeerClock(false);
+    request->hier.stopPeerClock();
```

Comment:
We want to pair start/stop calls, right? This stop call appears to _duplicate_ the stop call in retryOrBail() for cases where the server connection was opened, closed, and retryOrBail() was called. Same concern for this stop call and the stop call in sendError(). If I am wrong, let's add a C++ comment to explain why all these stop calls are mutually exclusive.

If accurate pairing is very difficult, we can make stopPeering() conditional, but that has its own drawbacks, so I would rather not cross that bridge unless we really have to.

Reply:
We cannot remove this call because it covers the case when retryOrBail() is not called (finishWritingAndDelete() handles the 'non-retriable connection closure'). On the other hand, serverClosed() may be called after retryOrBail(). The third stopPeering() call in sendError() is also needed, for example on the usePinned() path when retryOrBail() is not going to be called. It would be tempting to resolve this complexity by moving stopPeering() into the TunnelStateData() destructor, but in this case the peering time would be counted until all data has been written to the client (especially when the client is slow) and the server connection had been closed.
I am not sure that we have an elegant solution here. For now, I added a 'running()' check into stopPeering() to allow these multiple calls.

Reply:
I think we can simply this by adding a precondition to stopPeering() - whether the server connection is till open. I'll adjust accordingly.

Reply:
I examined all stopPeering() callers and think that it is possible to avoid them being called twice:
* serverClosed(): tunnelServerClosed() callback is assigned after retries are forbidden (i.e., no retryOrBail()). sendError() is not possible either: it is called from a) retryOrBail(), b) noteDestinationsEnd() - we have transporting() true already, c) usePinned(), which is called before tunnelServerClosed() callback is assigned.
* retryOrBail() and sendError(). Since retryOrBail() calls sendError() but sendError() does not lead to retryOrBail(), it is possible to move stopPeering() below in retryOrBail(), basically reverting your change [request](https://github.com/measurement-factory/squid/pull/249/files/8688bb5bff41c35167c0b3c4f7bd0a461612d309#r1540168016). Though you argued that  it is the best place for stopPeering() in retryOrBail(), I think it is better to sacrifice a little this code clarity but ensure calling stopPeering() only once.
* Also I think we should have an emergency stopPeering() in the TunnelStateData destructor.

Done (8092565).
I added an assure() that stopPeering() is called once - do we need some some explanation in the code why all three stopPeering() are required and all of them are unique?

Reply:
> I added an assure() that stopPeering() is called once 

I had to remove this check because our tests [revealed](https://github.com/measurement-factory/squid/actions/runs/8722360668/job/23928055369?pr=249) a problem: on the switching protocol path FwdState may be destroyed after TunnelStateData. We could add a flag checking that TunnelStateData called stopPeering() before destruction, but this looks excessive.

Reply:
> I examined all stopPeering() callers and think that it is possible to avoid them being called twice [in the current code]: [... complex proof follows ...]

I am happy to assume that the above conclusion is correct. Unfortunately, I am pretty sure that even small code modifications may invalidate the reasoning behind that conclusion; there are no invariants that will uphold all the current relationships between sendError(), retryOrBail(), noteDesinationsEnd(), and tunnelServerClosed(). And we do plan some significant code modifications in this area...

You have also demonstrated that current code requires stopPeering() calls from multiple same-task locations.

The combination tells me that a reliable solution requires introduction of a helper object. I bet we can keep that object very simple:

```C++
/// Eliminates excessive Stopwatch resume() and pause() calls in a task with
/// multiple code locations that resume (or pause) a stopwatch. Ideally, there
/// would be just one such location per action (e.g., constructor resumes while
/// destructor pauses), but current code idiosyncrasies necessitate this state.
class ActivityTimer
{
public:
    // TODO: Call timer.resume() here if possible, eliminating `resumed` and
    // adjusting class description to be exclusively about pause().
    ActivityTimer(Stopwatch &w): timer(w) {}

    ~ActivityTimer() { stop(); }

    start()
    {
        if (!resumed) {
            timer.resume();
            resumed = true;
        }
    }

    stop()
    {
        if (!paused) {
            timer.pause();
            paused = true;
        }
    }

private:
    Stopwatch &timer;

    // Do not be tempted to rely on timer.ran(): We are eliminating excessive
    // calls within a single task (e.g., an AsyncJob) while the timer (and its
    // ran() state) may be shared/affected by multiple concurrent tasks.
    bool resumed = false;
    bool paused = false;
};
```

If it is difficult to supply the right Stopwatch reference to the constructor in some cases, then let's refactor to supply a timer reference to start() and stop(). The latter would address Stopwatch object lifetime concerns.